File: Napoleon.ce/Core/kki/kkk/Android/modules/base/src/main/java/com/grsoft/napoleon/modules/print/util/Dig2Str.java
package com.grsoft.napoleon.modules.print.util;

public class Dig2Str {
	protected static String[] digSet = {" один", " два", " три", " четыре", " пять", " шесть", " семь", " восемь",
		  " девять", " десять", " одиннадцать", " двенадцать", " тринадцать",
		  " четырнадцать", " пятнадцать", " шестнадцать", " семнадцать",
		  " восемнадцать", " девятнадцать"};
	protected static String[] decDigSet = {" двадцать", " тридцать", " сорок", " пятьдесят", " шестьдесят",
		  " семьдесят", " восемьдесят", " девяносто"};
	protected static String[] hunDigSet =	{" сто", " двести", " триста", " четыреста", " пятьсот", " шестьсот",
		  " семьсот", " восемьсот", " девятьсот"};
	protected static String[] othDig = {"", " тысяч", " миллион", " миллиард"};
	protected static String[] shortOth = {"", " тыс", " млн", " млрд" };
	protected static String[] firstRest =	{"", "а", "и"};
	protected static String[] lastRest = {"ов", "", "а"};
	protected static String[] shortRest =	{".", ".", "." };


	private static String addRest( long val, String str, String base, String[] restSet)
	{
		  int rVal = (int)(val%100);
		  str += base;
		
		  if( rVal > 10 && rVal < 20 )
		  {
		     str += restSet[0];
		     return str;
		  }
		
		  rVal %= 10;
		  switch ( rVal )
		  {
		     case 1:
		        str += restSet[1];
		        return str;
		     case 2:
		     case 3:
		     case 4:
		        str += restSet[2];
		        return str;
		  }
		  str += restSet[0];
		  
		  return str;
	}


	private static String conv1000( int val, int step, String str) {
	  int rest = val % 100;
	  val /= 100;
	
	  if ( val > 0 ) 
		  str = hunDigSet[val-1];
	  else 
		  str = "";
	
	  if ( rest == 0 ) 
		  return str;
	  
	  if ( rest < 20 ) {
	     if ( step == 1 )
	        switch ( rest ) {
	           case 1:
	              str += " одна";
	              return str;
	           case 2:
	              str += " две";
	              return str;
	        }
	     str += digSet[rest-1];
	  } else {
	     str += decDigSet[(rest/10)-2];
	     if ( rest % 10  > 0) {
	        rest %= 10;
	        if ( step == 1 )
	           switch ( rest ) {
	              case 1:
	                 str += " одна";
	                 return str;
	              case 2:
	                 str += " две";
	              return str;
	           }
	        str += digSet[rest-1];
	    }
	  }
	  
	  return str;
	}

	public static String digToText(long dig)
	{
		int step = 0;
		long lastDig;
		String result = new String();
		
		if ( dig == 0 ) 
			return " ноль";
		
		do
		{
			int rest = (int)(dig%1000);
		    lastDig = dig;
		    dig /= 1000;
		    
		    if ( rest != 0 ) {
		    	String curDig = conv1000(rest, step, "");
		
		        if ( step != 0 )
		           curDig = addRest( lastDig, curDig, othDig[step], (step == 1) ? firstRest : lastRest );
		
		        curDig += result;
		        result = curDig;
		    }
		    
		    step++;
		} while ( dig > 0 );
		
		return result;
	} 
}

File: Napoleon.ce/Core/Napoleon.sql/Pda/Core/Dig2Str.cpp
/*
 * Copyright (C), 2007-2008, Денис Мосягин
 *
 * Реализация функций накладной
 *
 *  ert   17/05/2008   creating
 */
#include "stdafx.h"
#include "DoPrint.h"

wchar_t *digSet[] =
{
  L" один", L" два", L" три", L" четыре", L" пять", L" шесть", L" семь", L" восемь",
  L" девять", L" десять", L" одиннадцать", L" двенадцать", L" тринадцать",
  L" четырнадцать", L" пятнадцать", L" шестнадцать", L" семнадцать",
  L" восемнадцать", L" девятнадцать"
};

wchar_t *DecDigSet[] =
{
  L" двадцать", L" тридцать", L" сорок", L" пятьдесят", L" шестьдесят",
  L" семьдесят", L" восемьдесят", L" девяносто"
};

wchar_t *HunDigSet[] =
{
  L" сто", L" двести", L" триста", L" четыреста", L" пятьсот", L" шестьсот",
  L" семьсот", L" восемьсот", L" девятьсот"
};

wchar_t *OthDig[] =
{
  L"",
  L" тысяч",
  L" миллион",
  L" миллиард"
};

wchar_t *ShortOth[]=
{
  L"",
  L" тыс",
  L" млн",
  L" млрд"
};

const wchar_t *FirstRest[] =
{
  L"",
  L"а",
  L"и"
};

const wchar_t *LastRest[] =
{
  L"ов",
  L"",
  L"а"
};

const wchar_t *ShortRest[] =
{
  L".",
  L".",
  L"."
};

static void Conv1000( int val, int step, wchar_t *str);

void AddRest( long val, std::wstring *str, const wchar_t *base, const wchar_t *restSet[] )
{
  int rVal = (int)(val%100);
  (*str) += base;

  if( rVal > 10 && rVal < 20 )
  {
     (*str) += restSet[0];
     return;
  }

  rVal %= 10;
  switch ( rVal )
  {
     case 1:
        (*str) += restSet[1];
        return;
     case 2:
     case 3:
     case 4:
        (*str) += restSet[2];
        return;
  }
  (*str) += restSet[0];
}


static void Conv1000( int val, int step, std::wstring *str)
{
  int rest = val % 100;
  val /= 100;

  if ( val ) (*str) = HunDigSet[val-1];
  else str->clear();

  if ( !rest ) return;
  if ( rest < 20 )
  {
     if ( step == 1 )
        switch ( rest )
        {
           case 1:
              (*str) += L" одна";
              return;
           case 2:
              (*str) += L" две";
              return ;
        }
     (*str) += digSet[rest-1];
  } else
  {
     (*str) += DecDigSet[(rest/10)-2];
     if ( rest % 10 )
     {
        rest %= 10;
        if ( step == 1 )
           switch ( rest )
           {
              case 1:
                 (*str) += L" одна";
                 return;
              case 2:
                 (*str) += L" две";
              return ;
           }
        (*str) += digSet[rest-1];
    }
  }
}

void DigToText(std::wstring *str, DWORD dig)
{
  int step = 0;
  long lastDig;
  if ( dig == 0 )
  {
     *str = L" ноль";
     return;
  }

  str->clear();
  do
  {
     int rest = (int)(dig%1000);
     lastDig = dig;
     dig /= 1000;
    
     if ( rest != 0 )
     {
        std::wstring curDig;
        Conv1000(rest, step, &curDig);

        if ( step != 0 )
           AddRest( lastDig, &curDig, OthDig[step], (step == 1) ? FirstRest : LastRest );

        curDig += (*str);
        *str = curDig;
     }
     step++;
  } while ( dig > 0 );
}

